File: wi-seong-cheol/PGS/0x06 Queue/두 큐 합 같게 만들기.swift
//
//  두 큐 합 같게 만들기.swift
//  wi-seong-cheol
//
//  Created by wi_seong on 2022/12/16.
//

// 시간복잡도: O(n)
// 최악시간: 600,000
// 난이도: Level 2
// Timer: 32m
// Url: https://school.programmers.co.kr/learn/courses/30/lessons/118667?language=swift

import Foundation

func solution(_ queue1:[Int], _ queue2:[Int]) -> Int {
    var q1 = queue1
    var q2 = queue2
    var sum1 = q1.reduce(0, +)
    var sum2 = q2.reduce(0, +)
    var pointer1 = 0
    var pointer2 = 0
    for i in 0..<600010 {
        if sum1 == sum2 {
            return i
        }
        if sum1 < sum2 {
            q1.append(q2[pointer2])
            sum1 += q2[pointer2]
            sum2 -= q2[pointer2]
            pointer2 += 1
        } else {
            q2.append(q1[pointer1])
            sum2 += q1[pointer1]
            sum1 -= q1[pointer1]
            pointer1 += 1
        }
    }
    return -1
}

File: hall_of_fame/wi-seong-cheol/PGS/스킬 체크 테스트/Level 2/두 큐 합 같게 만들기.cpp
//
//  두 큐 합 같게 만들기.cpp
//  wi-seong-cheol
//
//  Created by wi_seong on 2023/03/02.
//

// 시간복잡도: O(2L)
// 최악시간: 600,000
// 난이도: Level 2
// Timer: 30m
// Url: https://school.programmers.co.kr/learn/courses/30/lessons/118667

#include <string>
#include <vector>
#include <queue>

using namespace std;
#define ll long long

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = 0;

    ll sum1, sum2;
    deque<int> q1, q2;
    sum1 = sum2 = 0;
    for(int q: queue1) {
        sum1 += q;
        q1.push_back(q);
    }
    for(int q: queue2) {
        sum2 += q;
        q2.push_back(q);
    }
    if((sum1 + sum2) % 2 == 1) return -1;
    int t = q1.size() + q2.size() + 2;
    while(sum1 != sum2) {
        if(answer > t + 2) return -1;
        else if(sum1 > sum2) {
            q2.push_back(q1.front());
            sum2 += q1.front();
            sum1 -= q1.front();
            q1.pop_front();
        } else {
            q1.push_back(q2.front());
            sum1 += q2.front();
            sum2 -= q2.front();
            q2.pop_front();
        }
        answer++;
    }

    return answer;
}
